rotkehlchen/chain/substrate/manager.py

```diff
@@ -0,0 +1,555 @@
+
+    def _get_last_block(self, node_interface: NodeInterface) -> BlockNumber:
+        """Return the chain height.
+        """
+        log.debug(f'{self.chain} querying last block', url=node_interface.url)
+        try:
+            last_block = node_interface.get_block_number(
+                block_hash=node_interface.get_chain_head(),
+            )
+        except requests.exceptions.RequestException as e:
+            raise RemoteError(
+                f'{self.chain} could not connect node at endpoint: {node_interface.url}. '
+                f'Connection error: {str(e)}.',
+            ) from e
+        except SubstrateRequestException as e:
+            raise RemoteError(
+                f'{self.chain} failed to request the last block: {str(e)}.',
+            ) from e
+
+        log.debug(f'{self.chain} last block', last_block=last_block)
+        return BlockNumber(last_block)
+
+    def _get_nodes_call_order(self) -> NodesCallOrder:
+        """Return a list of maps between a node and its attributes sorted by
+        preference.
+
+        TODO: once "own node" is introduced, refactor prioritizing it over the
+        rest.
+
+        NB: currently preference is based on how close they are to the chain
+        height requested via Subscan API, the higher 'weight_block' the better.
+        """
+        return sorted(
+            cast(Iterable, self.available_node_attributes_map.items()),
+            key=lambda item: -item[1].weight_block,
+        )
+
+    def _get_node_interface(self, node: NodeName) -> NodeInterface:
+        """Get an instance of SubstrateInterface.
+
+        May raise (most common):
+        - ValueError: during SubstrateInterface instantiation.
+        - FileNotFound: via `load_type_registry_preset()` if it doesn't exist
+        a preset file for the given `type_registry_preset` argument.
+
+        SubstrateInterface class:
+        https://github.com/polkascan/py-substrate-interface/blob/master/substrateinterface/base.py
+
+        Scale coded load_type_registry_preset:
+        https://github.com/polkascan/py-scale-codec/blob/master/scalecodec/type_registry/__init__.py  # noqa: E501
+        """
+        si_attributes = self.chain.get_substrate_interface_attributes()
+        try:
+            node_interface = SubstrateInterface(
+                url=node.endpoint(),
+                type_registry_preset=si_attributes.type_registry_preset,
+            )
+        except requests.exceptions.RequestException as e:
+            raise RemoteError(
+                f'{self.chain} could not connect to {node} node at '
+                f'endpoint: {node.endpoint()}. Connection error: {str(e)}.',
+            ) from e
+        except (FileNotFoundError, ValueError, TypeError) as e:
+            raise AttributeError('Invalid SubstrateInterface instantiation') from e
+
+        return node_interface
+
+    def _request_chain_api(self, endpoint: Literal['metadata']) -> Response:
```

Comment:
what other endpoints will we use in the future? Because if we won't use any perhaps this abstraction is not needed?

Reply:
The API is quite complete. Many rpc calls <SubstrateInterface> provides us are there, besides [staking](https://docs.api.subscan.io/#staking) per account, [price history,](https://docs.api.subscan.io/#price) etc. This is one of the reasons I've abstracted it.